### PR TITLE
[libc] at_quick_exit function implemented

### PR DESCRIPTION
**Pull request by aaryanshukla**
- added at_quick_exit function 
- used helper file exit_handler which reuses code from atexit
- atexit now calls helper functions from exit_handler
- test cases and dependencies are added

### CODE REVIEW

File: libc/config/linux/x86_64/entrypoints.txt [hunk 1]
```diff
@@ -751,6 +750,9 @@ if(LLVM_LIBC_FULL_BUILD)
     libc.src.stdlib.abort
     libc.src.stdlib.atexit
     libc.src.stdlib.exit
+    libc.src.stdlib.quick_exit
+    libc.src.stdlib.at_quick_exit
```
Comment (RoseZhang03) [inline libc/config/linux/x86_64/entrypoints.txt:754]: alphabetical order, at_quick_exit goes before atexit, quick_exit goes after getenv

File: libc/spec/stdc.td [hunk 2]
```diff
@@ -1089,6 +1089,7 @@ def StdC : StandardSpec<"stdc"> {
           FunctionSpec<"exit", RetValSpec<NoReturn>, [ArgSpec<IntType>]>,
           FunctionSpec<"atexit", RetValSpec<IntType>, [ArgSpec<AtexitHandlerT>]>,
           FunctionSpec<"quick_exit", RetValSpec<NoReturn>, [ArgSpec<IntType>]>,
+          FunctionSpec<"at_quick_exit", RetValSpec<IntType>, [ArgSpec<IntType>]>,
```
Comment (RoseZhang03) [inline libc/spec/stdc.td:1092]: alphabetical order

File: libc/src/stdlib/at_quick_exit.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
+//===-- Implementation of at_quick_exit -----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/stdlib/at_quick_exit.h"
+#include "src/__support/CPP/mutex.h" // lock_guard
```
Comment (RoseZhang03) [inline libc/src/stdlib/at_quick_exit.cpp:10]: Some includes are not directly used and can be removed, check VSCode problems tab

File: libc/src/stdlib/atexit.cpp [hunk 4]
```diff
@@ -12,91 +12,29 @@
 #include "src/__support/common.h"
 #include "src/__support/fixedvector.h"
 #include "src/__support/threads/mutex.h"
```
Comment (RoseZhang03) [inline libc/src/stdlib/atexit.cpp:14]: remove unneeded includes

File: libc/src/stdlib/atexit.cpp [hunk 5]
```diff
@@ -12,91 +12,29 @@
...
 
 LLVM_LIBC_FUNCTION(int, atexit, (StdCAtExitCallback * callback)) {
   return add_atexit_unit(
+      atexit_callbacks,
       {&stdc_at_exit_func, reinterpret_cast<void *>(callback)});
 }
 
 } // namespace LIBC_NAMESPACE
+
+
```
Comment (RoseZhang03) [inline libc/src/stdlib/atexit.cpp:40]: too many extra lines at end

File: libc/src/stdlib/exit_handler.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
+//===--- Implementation of exit_handler------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/stdlib/exit_handler.h"
```
Comment (RoseZhang03) [inline libc/src/stdlib/exit_handler.cpp:9]: unused includes

File: libc/src/stdlib/exit_handler.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+namespace LIBC_NAMESPACE {
+
+using AtExitCallback = void(void *);
+using StdCAtExitCallback = void(void);
+constexpr size_t CALLBACK_LIST_SIZE_FOR_TESTS = 1024;
+
+struct AtExitUnit {
+  AtExitCallback *callback = nullptr;
+  void *payload = nullptr;
+  constexpr AtExitUnit() = default;
```
Comment (RoseZhang03) [inline libc/src/stdlib/exit_handler.h:27]: add LIBC_INLINE tag in front of the two constexpr lines

File: libc/test/src/stdlib/at_quick_exit_test.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+TEST(LlvmLibcAtQuickExit, HandlerCallsAtQuickExit) {
+  auto test = [] {
+    LIBC_NAMESPACE::at_quick_exit(+[] {
+      LIBC_NAMESPACE::at_quick_exit(+[] { LIBC_NAMESPACE::quick_exit(1); });
+    });
+    LIBC_NAMESPACE::quick_exit(0);
+  };
+  EXPECT_EXITS(test, 1);
+}
```
Comment (RoseZhang03) [inline libc/test/src/stdlib/at_quick_exit_test.cpp:90]: add line at the end

**Review by RoseZhang03 — COMMENTED** [review]
LGTM!

File: libc/test/src/stdlib/CMakeLists.txt
Comment (RoseZhang03) [inline libc/test/src/stdlib/CMakeLists.txt:1]: For atexit_test: CPP.utility can be removed (double check includes on atexit_test.cpp)

Comment (RoseZhang03) [reply libc/test/src/stdlib/CMakeLists.txt:1]: Also double check at_quick_exit.cpp just in case

File: libc/src/stdlib/CMakeLists.txt [hunk 9]
```diff
@@ -432,6 +433,28 @@ add_entrypoint_object(
     libc.src.__support.threads.mutex
 )
 
+add_entrypoint_object(
+  atexit
+  SRCS
+    atexit.cpp
+  HDRS
+    atexit.h
+  CXX_STANDARD
```
Comment (michaelrj-google) [inline libc/src/stdlib/CMakeLists.txt:442]: this doesn't have anything constinit in it anymore, so it doesn't need this `CXX_STANDARD` annotation.

**Review by michaelrj-google — COMMENTED** [review]
Overall LGTM with some minor changes

**Review by RoseZhang03 — COMMENTED** [review]
LGTM! (for real this time)

**Review by michaelrj-google — APPROVED** [review]
LGTM, good job!

File: libc/src/stdlib/at_quick_exit.h [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC_STDLIB_AT_QUICK_EXIT_H
+#define LLVM_LIBC_SRC_STDLIB_AT_QUICK_EXIT_H
+
+namespace LIBC_NAMESPACE {
+
+int at_quick_exit(void (*function)());
```
Comment (lntue) [inline libc/src/stdlib/at_quick_exit.h:14]: The function signatures in `at_quick_exit.h` and `at_quick_exit.cpp` should be identical.

File: libc/hdr/types/atexithandler_t.h [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_HDR_ATEXITHANDLER_T_H
+#define LLVM_LIBC_HDR_ATEXITHANDLER_T_H
+
+#ifdef LIBC_FULL_BUILD
+
+#include "include/llvm-libc-types/atexithandler_t.h"
+
```
Comment (michaelrj-google) [inline libc/hdr/types/atexithandler_t.h:15]: there needs to be an `#else` here.

File: libc/src/stdlib/atexit.h [hunk 12]
```diff
@@ -8,13 +8,8 @@
 
 #ifndef LLVM_LIBC_SRC_STDLIB_ATEXIT_H
 #define LLVM_LIBC_SRC_STDLIB_ATEXIT_H
-
-#include <stddef.h> // For size_t
```
Comment (michaelrj-google) [inline libc/src/stdlib/atexit.h:12]: this should include the proxy header for `__atexithandler_t` and use it in the function prototype

File: libc/src/stdlib/exit_handler.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+#include "src/__support/CPP/mutex.h" // lock_guard
+#include "src/__support/blockstore.h"
+#include "src/__support/common.h"
+#include "src/__support/fixedvector.h"
+#include "src/__support/threads/mutex.h"
+
+namespace LIBC_NAMESPACE {
+
+using AtExitCallback = void(void *);
```
Comment (michaelrj-google) [inline libc/src/stdlib/exit_handler.h:20]: you should include `__atexithandler_t` and change this using to be `using AtExitCallback = __atexithandler_t`

File: libc/src/stdlib/at_quick_exit.h (hunk 10 above)
Comment (michaelrj-google) [inline libc/src/stdlib/at_quick_exit.h:14]: this should include the proxy header for `__atexithandler_t` and use it in the function prototype

**Review by michaelrj-google — COMMENTED** [review]
needs a couple more changes with the proxy header update

**Review by michaelrj-google — APPROVED** [review]
LGTM, will land tomorrow morning

**Comment by zeroomega** [comment]
@aaryanshukla We discovered a cmake configuration error when building the baremetal runtime `runtimes-riscv32-unknown-elf` on builder https://lab.llvm.org/buildbot/#/builders/98/builds/39050 after this change is landed. Error message:

```
-- Configuring done (2.5s)
CMake Error at /var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/libc/cmake/modules/LLVMLibCObjectRules.cmake:90 (add_dependencies):
  The dependency target "libc.src.__support.threads.mutex" of target
  "libc.src.stdlib.exit_handler" does not exist.
Call Stack (most recent call first):
  /var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/libc/cmake/modules/LLVMLibCFlagRules.cmake:158 (create_object_library)
  /var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/libc/cmake/modules/LLVMLibCFlagRules.cmake:158 (cmake_language)
  /var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/libc/cmake/modules/LLVMLibCFlagRules.cmake:251 (expand_flags_for_target)
  /var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/libc/cmake/modules/LLVMLibCObjectRules.cmake:122 (add_target_with_flags)
  /var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/libc/src/stdlib/CMakeLists.txt:419 (add_object_library)
-- Generating done (0.4s)
CMake Generate step failed.  Build files cannot be regenerated correctly.
```
Could you take a look please?

**Comment by michaelrj-google** [comment]
We are looking into it, hopefully have a fix soon!

**Pull request merged**